salt/templates/virt/libvirt_domain.jinja

```diff
@@ -1,32 +1,219 @@
+                             <cache id='{{ e }}' {%- for atr, val in atrs.items() %} {{ atr }}='{{ val }}' {%- endfor %} />
+                        {%- elif e is not number %}
+                             {%- for atr, val in atrs.items() %}
+                             <monitor level='{{ val }}' vcpus='{{ atr }}'/>
+                             {%- endfor %}
+                        {%- endif %}
+                    {%- endfor %}
+                </cachetune>
+                {%- endfor %}
+            {%- endif %}
+            {%- if cpu.tuning.memorytune %}
+                {%- for vcpus, nodes in cpu.tuning.memorytune.items() %}
+                <memorytune vcpus='{{ vcpus}}'>
+                    {%- for id, bandwidth in nodes.items() %}
+                         <node id='{{ id }}' bandwidth='{{ bandwidth }}'/>
+                    {%- endfor %}
+                </memorytune>
+                {%- endfor %}
+            {%- endif %}
+         </cputune>
+        {%- endif %}
         {%- if mem.max %}
-        <maxMemory {{ mem.slots }} unit='KiB'> {{ mem.max }}</maxMemory>
+        <maxMemory {{ opt_attribute(mem, 'slots') }} unit='KiB'>{{ to_kib(mem.max) }}</maxMemory>
         {%- endif %}
         {%- if mem.boot %}
-        <memory unit='KiB'>{{ mem.boot }}</memory>
+        <memory unit='KiB'>{{ to_kib(mem.boot) }}</memory>
         {%- endif %}
         {%- if mem.current %}
-        <currentMemory unit='KiB'>{{ mem.current }}</currentMemory>
+        <currentMemory unit='KiB'>{{ to_kib(mem.current) }}</currentMemory>
         {%- endif %}
         {%- if mem %}
         <memtune>
             {%- if 'hard_limit' in mem and mem.hard_limit %}
-                <hard_limit unit="KiB">{{ mem.hard_limit }}</hard_limit>
+                <hard_limit unit="KiB">{{ to_kib(mem.hard_limit) }}</hard_limit>
             {%- endif %}
             {%- if 'soft_limit' in mem and mem.soft_limit %}
-                <soft_limit unit="KiB">{{ mem.soft_limit }}</soft_limit>
+                <soft_limit unit="KiB">{{ to_kib(mem.soft_limit) }}</soft_limit>
             {%- endif %}
             {%- if 'swap_hard_limit' in mem and mem.swap_hard_limit %}
-                <swap_hard_limit unit="KiB">{{ mem.swap_hard_limit }}</swap_hard_limit>
+                <swap_hard_limit unit="KiB">{{ to_kib(mem.swap_hard_limit) }}</swap_hard_limit>
             {%- endif %}
             {%- if 'min_guarantee' in mem and mem.min_guarantee %}
-                <min_guarantee unit="KiB">{{ mem.min_guarantee }}</min_guarantee>
+                <min_guarantee unit="KiB">{{ to_kib(mem.min_guarantee) }}</min_guarantee>
             {%- endif %}
         </memtune>
         {%- endif %}
+        {%- if numatune %}
+        <numatune>
+            {%- if 'memory' in numatune and numatune.memory %}
+                <memory mode='{{ numatune.memory.mode }}'
+                    {%- if numatune.memory.nodeset %}nodeset='{{ numatune.memory.nodeset }}'{%- endif %}
```

Comment:
```suggestion
                    {%- if numatune.memory.nodeset %} nodeset='{{ numatune.memory.nodeset }}'{%- endif %}
```

There is a missing white space before the `nodeset` attribute that makes the XML invalid